Wrong collection type in Set assertion

#### Summary

The assertion error displays the wrong collection type, as noticed by @Patouche.

#### Example

```java
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class AssertJTests {

    @Test
    void iterable_bad_name() {
        final Set<String> actual = Stream.of("toto", "tata").collect(Collectors.toSet());
        Assertions.assertThat(actual).contains("tutu");
    }
}
```

The error is as follows:
```java
java.lang.AssertionError: 
Expecting ArrayList:
 <["toto", "tata"]>
to contain:
 <["tutu"]>
but could not find the following element(s):
 <["tutu"]>
```

`ArrayList` should not be displayed in the error message.

--- a/src/main/java/org/assertj/core/error/GroupTypeDescription.java
+++ b/src/main/java/org/assertj/core/error/GroupTypeDescription.java
@@ -42,12 +42,21 @@ public class GroupTypeDescription {
    * @return the created GroupTypeDescription object
    */
   public static GroupTypeDescription getGroupTypeDescription(Object actual) {
+    return getGroupTypeDescription(actual.getClass());
+  }
 
-    Class<?> clazz = actual.getClass();
+  /**
+   * Creates a new <code>{@link GroupTypeDescription}</code> for a group of elements.
+   *
+   * @param clazz the class for the group of elements.
+   * @return the created GroupTypeDescription object
+   */
+  public static GroupTypeDescription getGroupTypeDescription(Class<?> clazz) {
+
     if (Thread.currentThread().getStackTrace()[SPLITERATORS_CLASS_STACK_TRACE_NUM].getClassName().contains("Spliterators"))
       return new GroupTypeDescription("spliterator characteristics", "characteristics");
 
-    if (actual instanceof Map) return new GroupTypeDescription("map", "map entries");
+    if (Map.class.isAssignableFrom(clazz)) return new GroupTypeDescription("map", "map entries");
 
     if (clazz.isArray())
       return new GroupTypeDescription(clazz.getSimpleName(), clazz.getComponentType().getSimpleName().toLowerCase() + "(s)");

--- a/src/main/java/org/assertj/core/error/ShouldContain.java
+++ b/src/main/java/org/assertj/core/error/ShouldContain.java
@@ -41,7 +41,21 @@ public class ShouldContain extends BasicErrorMessageFactory {
    */
   public static ErrorMessageFactory shouldContain(Object actual, Object expected, Object notFound,
                                                   ComparisonStrategy comparisonStrategy) {
-    GroupTypeDescription groupTypeDescription = getGroupTypeDescription(actual);
+    return shouldContain(actual.getClass(), actual, expected, notFound, comparisonStrategy);
+  }
+
+  /**
+   * Creates a new <code>{@link ShouldContain}</code>.
+   * @param clazz the actual value class in the failed assertion.
+   * @param actual the actual value in the failed assertion.
+   * @param expected values expected to be in {@code actual}.
+   * @param notFound the values in {@code expected} not found in {@code actual}.
+   * @param comparisonStrategy the {@link ComparisonStrategy} used to evaluate assertion.
+   * @return the created {@code ErrorMessageFactory}.
+   */
+  public static ErrorMessageFactory shouldContain(Class<?> clazz, Object actual, Object expected, Object notFound,
+                                                  ComparisonStrategy comparisonStrategy) {
+    GroupTypeDescription groupTypeDescription = getGroupTypeDescription(clazz);
     return new ShouldContain(actual, expected, notFound, comparisonStrategy, groupTypeDescription);
   }
 

--- a/src/main/java/org/assertj/core/internal/Iterables.java
+++ b/src/main/java/org/assertj/core/internal/Iterables.java
@@ -342,15 +342,16 @@ public class Iterables {
     final List<?> actualAsList = newArrayList(actual);
     if (commonCheckThatIterableAssertionSucceeds(info, actualAsList, values)) return;
     // check for elements in values that are missing in actual.
-    assertIterableContainsGivenValues(actualAsList, values, info);
+    assertIterableContainsGivenValues(actual.getClass(), actualAsList, values, info);
   }
 
-  private void assertIterableContainsGivenValues(Iterable<?> actual, Object[] values, AssertionInfo info) {
+  private void assertIterableContainsGivenValues(@SuppressWarnings("unchecked") Class<? extends Iterable> clazz,
+                                                 Iterable<?> actual, Object[] values, AssertionInfo info) {
     Set<Object> notFound = stream(values).filter(value -> !iterableContains(actual, value))
                                          .collect(toCollection(LinkedHashSet::new));
     if (notFound.isEmpty())
       return;
-    throw failures.failure(info, shouldContain(actual, values, notFound, comparisonStrategy));
+    throw failures.failure(info, shouldContain(clazz, actual, values, notFound, comparisonStrategy));
   }
 
   private boolean iterableContains(Iterable<?> actual, Object value) {
@@ -1099,10 +1100,11 @@ public class Iterables {
    *           {@code Iterable}, in any order.
    */
   public void assertContainsAll(AssertionInfo info, Iterable<?> actual, Iterable<?> other) {
+    final List<?> actualAsList = newArrayList(actual);
     checkIterableIsNotNull(other);
-    assertNotNull(info, actual);
+    assertNotNull(info, actualAsList);
     Object[] values = newArrayList(other).toArray();
-    assertIterableContainsGivenValues(actual, values, info);
+    assertIterableContainsGivenValues(actual.getClass(), actualAsList, values, info);
   }
 
   /**

--- a/src/test/java/org/assertj/core/error/GroupTypeDescription_getGroupTypeDescription_Test.java
+++ b/src/test/java/org/assertj/core/error/GroupTypeDescription_getGroupTypeDescription_Test.java
@@ -40,6 +40,16 @@ class GroupTypeDescription_getGroupTypeDescription_Test {
     assertThat(description.getElementTypeName()).isEqualTo(elementTypeName);
   }
 
+  @ParameterizedTest(name = "{0}: {1} {2} ")
+  @MethodSource("argumentsStreamProvider")
+  void should_return_group_description_from_class(Object obj, String groupTypeName, String elementTypeName) {
+    // WHEN
+    GroupTypeDescription description = getGroupTypeDescription(obj.getClass());
+    // THEN
+    assertThat(description.getGroupTypeName()).isEqualTo(groupTypeName);
+    assertThat(description.getElementTypeName()).isEqualTo(elementTypeName);
+  }
+
   private static Stream<Arguments> argumentsStreamProvider() {
     return Stream.of(Arguments.of(mapOf(entry("1", 2d)), "map", "map entries"),
                      Arguments.of(new int[] { 1, 2 }, "int[]", "int(s)"),

--- a/src/test/java/org/assertj/core/error/GroupTypeDescription_getGroupTypeDescription_Test.java
+++ b/src/test/java/org/assertj/core/error/GroupTypeDescription_getGroupTypeDescription_Test.java
@@ -12,12 +12,13 @@
  */
 package org.assertj.core.error;
 
-import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.entry;
+import static org.assertj.core.api.BDDAssertions.then;
 import static org.assertj.core.error.GroupTypeDescription.getGroupTypeDescription;
 import static org.assertj.core.test.Maps.mapOf;
 import static org.assertj.core.util.Lists.list;
 import static org.assertj.core.util.Sets.newLinkedHashSet;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
 
 import java.util.stream.Stream;
 
@@ -36,8 +37,8 @@ class GroupTypeDescription_getGroupTypeDescription_Test {
     // WHEN
     GroupTypeDescription description = getGroupTypeDescription(obj);
     // THEN
-    assertThat(description.getGroupTypeName()).isEqualTo(groupTypeName);
-    assertThat(description.getElementTypeName()).isEqualTo(elementTypeName);
+    then(description.getGroupTypeName()).isEqualTo(groupTypeName);
+    then(description.getElementTypeName()).isEqualTo(elementTypeName);
   }
 
   @ParameterizedTest(name = "{0}: {1} {2} ")
@@ -46,25 +47,25 @@ class GroupTypeDescription_getGroupTypeDescription_Test {
     // WHEN
     GroupTypeDescription description = getGroupTypeDescription(obj.getClass());
     // THEN
-    assertThat(description.getGroupTypeName()).isEqualTo(groupTypeName);
-    assertThat(description.getElementTypeName()).isEqualTo(elementTypeName);
+    then(description.getGroupTypeName()).isEqualTo(groupTypeName);
+    then(description.getElementTypeName()).isEqualTo(elementTypeName);
   }
 
   private static Stream<Arguments> argumentsStreamProvider() {
-    return Stream.of(Arguments.of(mapOf(entry("1", 2d)), "map", "map entries"),
-                     Arguments.of(new int[] { 1, 2 }, "int[]", "int(s)"),
-                     Arguments.of(new double[] { 1, 2 }, "double[]", "double(s)"),
-                     Arguments.of(new float[] { 1f, 2f }, "float[]", "float(s)"),
-                     Arguments.of(new byte[] { 1, 2 }, "byte[]", "byte(s)"),
-                     Arguments.of(new long[] { 1L, 2L }, "long[]", "long(s)"),
-                     Arguments.of(new boolean[] { true }, "boolean[]", "boolean(s)"),
-                     Arguments.of(new char[] { 'a', 'b' }, "char[]", "char(s)"),
-                     Arguments.of(new short[] { 1, 2 }, "short[]", "short(s)"),
-                     Arguments.of(new String[] { "a", "c" }, "String[]", "string(s)"),
-                     Arguments.of(new Object[] { "a", "c" }, "Object[]", "object(s)"),
-                     Arguments.of(new Jedi[] { new Jedi("Yoda", "green") }, "Jedi[]", "jedi(s)"),
-                     Arguments.of(list(1, 2), "ArrayList", "element(s)"),
-                     Arguments.of(newLinkedHashSet(1, 2), "LinkedHashSet", "element(s)"));
+    return Stream.of(arguments(mapOf(entry("1", 2d)), "map", "map entries"),
+                     arguments(new int[] { 1, 2 }, "int[]", "int(s)"),
+                     arguments(new double[] { 1, 2 }, "double[]", "double(s)"),
+                     arguments(new float[] { 1f, 2f }, "float[]", "float(s)"),
+                     arguments(new byte[] { 1, 2 }, "byte[]", "byte(s)"),
+                     arguments(new long[] { 1L, 2L }, "long[]", "long(s)"),
+                     arguments(new boolean[] { true }, "boolean[]", "boolean(s)"),
+                     arguments(new char[] { 'a', 'b' }, "char[]", "char(s)"),
+                     arguments(new short[] { 1, 2 }, "short[]", "short(s)"),
+                     arguments(new String[] { "a", "c" }, "String[]", "string(s)"),
+                     arguments(new Object[] { "a", "c" }, "Object[]", "object(s)"),
+                     arguments(new Jedi[] { new Jedi("Yoda", "green") }, "Jedi[]", "jedi(s)"),
+                     arguments(list(1, 2), "ArrayList", "element(s)"),
+                     arguments(newLinkedHashSet(1, 2), "LinkedHashSet", "element(s)"));
   }
 
 }

--- a/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContainsAll_Test.java
+++ b/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContainsAll_Test.java
@@ -18,7 +18,9 @@ import static org.assertj.core.api.Assertions.assertThatNullPointerException;
 import static org.assertj.core.api.Assertions.catchThrowable;
 import static org.assertj.core.error.ShouldContain.shouldContain;
 import static org.assertj.core.internal.ErrorMessages.iterableToLookForIsNull;
+import static org.assertj.core.internal.iterables.SinglyIterableFactory.createSinglyIterable;
 import static org.assertj.core.test.TestData.someInfo;
+import static org.assertj.core.util.Arrays.array;
 import static org.assertj.core.util.FailureMessages.actualIsNull;
 import static org.assertj.core.util.Lists.newArrayList;
 import static org.assertj.core.util.Sets.newLinkedHashSet;
@@ -54,6 +56,11 @@ class Iterables_assertContainsAll_Test extends IterablesBaseTest {
   @Test
   void should_pass_if_actual_contains_all_iterable_values_even_if_duplicated() {
     iterables.assertContainsAll(someInfo(), actual, newArrayList("Luke", "Luke"));
+  }
+
+  @Test
+  void should_pass_if_nonrestartable_actual_contains_given_values() {
+    iterables.assertContainsAll(someInfo(), createSinglyIterable(actual), newArrayList("Luke", "Luke"));
   }
 
   @Test

--- a/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java
+++ b/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java
@@ -12,10 +12,10 @@
  */
 package org.assertj.core.internal.iterables;
 
-import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
 import static org.assertj.core.api.Assertions.assertThatNullPointerException;
 import static org.assertj.core.api.Assertions.catchThrowable;
+import static org.assertj.core.api.BDDAssertions.then;
 import static org.assertj.core.error.ShouldContain.shouldContain;
 import static org.assertj.core.internal.ErrorMessages.valuesToLookForIsNull;
 import static org.assertj.core.internal.iterables.SinglyIterableFactory.createSinglyIterable;
@@ -101,24 +101,30 @@ class Iterables_assertContains_Test extends IterablesBaseTest {
 
   @Test
   void should_fail_if_actual_does_not_contain_values() {
+    // GIVEN
     AssertionInfo info = someInfo();
     Object[] expected = { "Han", "Luke" };
 
+    // WHEN
     Throwable error = catchThrowable(() -> iterables.assertContains(info, actual, expected));
 
-    assertThat(error).isInstanceOf(AssertionError.class);
+    // THEN
+    then(error).isInstanceOf(AssertionError.class);
     verify(failures).failure(info, shouldContain(actual, expected, newLinkedHashSet("Han")));
   }
 
   @Test
   void should_fail_with_the_right_actual_type() {
+    // GIVEN
     AssertionInfo info = someInfo();
-    Object[] expected = {"Han", "Luke"};
+    Object[] expected = { "Han", "Luke" };
     Set<String> actualSet = new HashSet<>(actual);
 
+    // WHEN
     Throwable error = catchThrowable(() -> iterables.assertContains(info, actualSet, expected));
 
-    assertThat(error).isInstanceOf(AssertionError.class).hasMessageContainingAll("Expecting HashSet:");
+    // THEN
+    then(error).isInstanceOf(AssertionError.class).hasMessageContainingAll("Expecting HashSet:");
     verify(failures).failure(info, shouldContain(HashSet.class, newArrayList(actualSet), expected, newLinkedHashSet("Han"), StandardComparisonStrategy.instance()));
   }
 
@@ -154,12 +160,15 @@ class Iterables_assertContains_Test extends IterablesBaseTest {
 
   @Test
   void should_fail_if_actual_does_not_contain_values_according_to_custom_comparison_strategy() {
+    // GIVEN
     AssertionInfo info = someInfo();
     Object[] expected = { "Han", "Luke" };
 
+    // WHEN
     Throwable error = catchThrowable(() -> iterablesWithCaseInsensitiveComparisonStrategy.assertContains(info, actual, expected));
 
-    assertThat(error).isInstanceOf(AssertionError.class);
+    // THEN
+    then(error).isInstanceOf(AssertionError.class);
     verify(failures).failure(info, shouldContain(actual, expected, newLinkedHashSet("Han"), comparisonStrategy));
   }
 

--- a/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java
+++ b/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java
@@ -14,7 +14,6 @@ package org.assertj.core.internal.iterables;
 
 import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
 import static org.assertj.core.api.Assertions.assertThatNullPointerException;
-import static org.assertj.core.api.Assertions.catchThrowable;
 import static org.assertj.core.api.BDDAssertions.then;
 import static org.assertj.core.error.ShouldContain.shouldContain;
 import static org.assertj.core.internal.ErrorMessages.valuesToLookForIsNull;
@@ -22,12 +21,12 @@ import static org.assertj.core.internal.iterables.SinglyIterableFactory.createSi
 import static org.assertj.core.test.ObjectArrays.emptyArray;
 import static org.assertj.core.test.TestData.someInfo;
 import static org.assertj.core.util.Arrays.array;
+import static org.assertj.core.util.AssertionsUtil.expectAssertionError;
 import static org.assertj.core.util.FailureMessages.actualIsNull;
 import static org.assertj.core.util.Lists.newArrayList;
 import static org.assertj.core.util.Sets.newLinkedHashSet;
 import static org.mockito.Mockito.verify;
 
-import java.util.Collection;
 import java.util.HashSet;
 import java.util.Set;
 
@@ -38,7 +37,7 @@ import org.assertj.core.internal.StandardComparisonStrategy;
 import org.junit.jupiter.api.Test;
 
 /**
- * Tests for <code>{@link Iterables#assertContains(AssertionInfo, Collection, Object[])}</code>.
+ * Tests for <code>{@link Iterables#assertContains(AssertionInfo, Iterable, Object[])}</code>.
  *
  * @author Alex Ruiz
  * @author Joel Costigliola
@@ -78,7 +77,7 @@ class Iterables_assertContains_Test extends IterablesBaseTest {
   }
 
   @Test
-  void should_pass_if_nonrestartable_actual_contains_given_values() {
+  void should_pass_if_non_restartable_actual_contains_given_values() {
     iterables.assertContains(someInfo(), createSinglyIterable(actual), array("Luke", "Yoda", "Leia"));
   }
 
@@ -104,12 +103,10 @@ class Iterables_assertContains_Test extends IterablesBaseTest {
     // GIVEN
     AssertionInfo info = someInfo();
     Object[] expected = { "Han", "Luke" };
-
     // WHEN
-    Throwable error = catchThrowable(() -> iterables.assertContains(info, actual, expected));
-
+    AssertionError error = expectAssertionError(() -> iterables.assertContains(info, actual, expected));
     // THEN
-    then(error).isInstanceOf(AssertionError.class);
+    then(error).hasMessageContaining("Expecting ArrayList:");
     verify(failures).failure(info, shouldContain(actual, expected, newLinkedHashSet("Han")));
   }
 
@@ -119,13 +116,12 @@ class Iterables_assertContains_Test extends IterablesBaseTest {
     AssertionInfo info = someInfo();
     Object[] expected = { "Han", "Luke" };
     Set<String> actualSet = new HashSet<>(actual);
-
     // WHEN
-    Throwable error = catchThrowable(() -> iterables.assertContains(info, actualSet, expected));
-
+    AssertionError error = expectAssertionError(() -> iterables.assertContains(info, actualSet, expected));
     // THEN
-    then(error).isInstanceOf(AssertionError.class).hasMessageContainingAll("Expecting HashSet:");
-    verify(failures).failure(info, shouldContain(HashSet.class, newArrayList(actualSet), expected, newLinkedHashSet("Han"), StandardComparisonStrategy.instance()));
+    then(error).hasMessageContaining("Expecting HashSet:");
+    verify(failures).failure(info, shouldContain(HashSet.class, newArrayList(actualSet), expected, newLinkedHashSet("Han"),
+                                                 StandardComparisonStrategy.instance()));
   }
 
   // ------------------------------------------------------------------------------------------------------------------
@@ -163,12 +159,11 @@ class Iterables_assertContains_Test extends IterablesBaseTest {
     // GIVEN
     AssertionInfo info = someInfo();
     Object[] expected = { "Han", "Luke" };
-
     // WHEN
-    Throwable error = catchThrowable(() -> iterablesWithCaseInsensitiveComparisonStrategy.assertContains(info, actual, expected));
-
+    AssertionError error = expectAssertionError(() -> iterablesWithCaseInsensitiveComparisonStrategy.assertContains(info, actual,
+                                                                                                                    expected));
     // THEN
-    then(error).isInstanceOf(AssertionError.class);
+    then(error).hasMessageContaining("Expecting ArrayList:");
     verify(failures).failure(info, shouldContain(actual, expected, newLinkedHashSet("Han"), comparisonStrategy));
   }
 

--- a/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java
+++ b/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java
@@ -34,6 +34,7 @@ import java.util.Set;
 import org.assertj.core.api.AssertionInfo;
 import org.assertj.core.internal.Iterables;
 import org.assertj.core.internal.IterablesBaseTest;
+import org.assertj.core.internal.StandardComparisonStrategy;
 import org.junit.jupiter.api.Test;
 
 /**
@@ -118,7 +119,7 @@ class Iterables_assertContains_Test extends IterablesBaseTest {
     Throwable error = catchThrowable(() -> iterables.assertContains(info, actualSet, expected));
 
     assertThat(error).isInstanceOf(AssertionError.class).hasMessageContainingAll("Expecting HashSet:");
-    verify(failures).failure(info, shouldContain(actualSet, expected, newLinkedHashSet("Han")));
+    verify(failures).failure(info, shouldContain(HashSet.class, newArrayList(actualSet), expected, newLinkedHashSet("Han"), StandardComparisonStrategy.instance()));
   }
 
   // ------------------------------------------------------------------------------------------------------------------

--- a/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java
+++ b/src/test/java/org/assertj/core/internal/iterables/Iterables_assertContains_Test.java
@@ -28,6 +28,8 @@ import static org.assertj.core.util.Sets.newLinkedHashSet;
 import static org.mockito.Mockito.verify;
 
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
 
 import org.assertj.core.api.AssertionInfo;
 import org.assertj.core.internal.Iterables;
@@ -105,6 +107,18 @@ class Iterables_assertContains_Test extends IterablesBaseTest {
 
     assertThat(error).isInstanceOf(AssertionError.class);
     verify(failures).failure(info, shouldContain(actual, expected, newLinkedHashSet("Han")));
+  }
+
+  @Test
+  void should_fail_with_the_right_actual_type() {
+    AssertionInfo info = someInfo();
+    Object[] expected = {"Han", "Luke"};
+    Set<String> actualSet = new HashSet<>(actual);
+
+    Throwable error = catchThrowable(() -> iterables.assertContains(info, actualSet, expected));
+
+    assertThat(error).isInstanceOf(AssertionError.class).hasMessageContainingAll("Expecting HashSet:");
+    verify(failures).failure(info, shouldContain(actualSet, expected, newLinkedHashSet("Han")));
   }
 
   // ------------------------------------------------------------------------------------------------------------------

